### ISSUE TITLE
One pixel shifted to the right?

### ISSUE THREAD

**Issue by landloafer** [AppleWin/AppleWin#555]
I just saw the title screen screen of prince of persia (DHGR) 
![apple_000](https://user-images.githubusercontent.com/39210397/39954334-3f8f5e0c-55f0-11e8-851c-990f0125051b.png)
with APPLEPC2.52 (dos). I noticed there is an extra line at the right most. But can't verify which one is correct.

**Comment by landloafer** [comment]
![applewin_screenshot_000000002](https://user-images.githubusercontent.com/39210397/39954467-a1e26ef8-55f2-11e8-9f76-bddeda0446ad.png)
1.27.2.0

**Comment by landloafer** [comment]
![applewin_screenshot_000000001](https://user-images.githubusercontent.com/39210397/39954479-cd212794-55f2-11e8-817b-a9e00ed7173a.png)
1.27.2.0

**Comment by tomcw** [comment]
Thanks for reporting this. Yes, it certainly looks like a bug introduced after 1.25.04 (ie. 1.26+).
I'll take a look at a fix.

**Comment by tomcw** [comment]
>While investigating this, I noticed there is a bug in AppleWin's NTSC.cpp, `updateVideoScannerHorzEOL()` ...

Fixed in 3e7cc36.

**Comment by landloafer** [comment]
Thanks!
Is it the whole screen shifted one pixel?

**Comment by tomcw** [comment]
>Is it the whole screen shifted one pixel?

Yes... but only DHGR, so in mixed mode there's a 1 pixel misalignment between DHGR (top 160 lines) & the bottom 32 lines of TEXT/TEXT80/GR/DGR etc.

Here's a zoomed display with the dash-line being the 159th DHGR line, and then TEXT80 below. It demonstrates the misalignment: (NB. screenshot from Sheldon's AppleWin-NTSC-05 build, but AppleWin 1.26/1.27 will be the same.)
![ntsc05-dhgr-mixed-zoomed](https://user-images.githubusercontent.com/6696896/40140608-48c3b348-594b-11e8-9fc5-038b4ea2892b.png)

**Comment by landloafer** [comment]
Interesting read! So anticipated to next release.

**Comment by Michaelangel007** [comment]
Tom, thanks for the fixing that. I guess I should have done more testing -- I didn't even think to try DHGR monochrome!

**Comment by tomcw** [comment]
Here's the HCOLOR vertical  line before the fix:
![aw1 27](https://user-images.githubusercontent.com/6696896/40515195-d1c93e70-5fa3-11e8-8686-5ab5a2c54996.PNG)

But now an HGR HCOLOR=2 vertical line at column 0 is only 1 DHGR pixel wide:
![aw1 27-fix](https://user-images.githubusercontent.com/6696896/40515200-d4194576-5fa3-11e8-98b1-7d34bb08b116.PNG)

Is this correct?
You'd need to check on a real TV or monitor... and definitely easier to verify on the monitor.

**Comment by tomcw** [comment]
NB. The DHGR 14-bit word is always shifted right by a dot.
But the HGR (expanded) 14-bit word is only shifted right by a dot when the byte's high bit is set.
So there's only this missing half HGR (or 1 DHGR) pixel for HGR column 0 when bit7=0.

**Comment by tomcw** [comment]
I'm going to close this issue, as I think it's correct now (including the HGR HCOLOR=2 characteristic).
Try [1.27.4](https://github.com/AppleWin/AppleWin/releases/download/v1.27.4.0/AppleWin1.27.4.0.zip) to test it.

**Comment by tomcw** [comment]
Re-opening, as I checked on real h/w (//e with a monochrome monitor) and 1.27.4 is still not correct:
- the HCOLOR=2 vertical line in column-0 is a full width pixel, not a half width / DHGR pixel.
  - and this aligns perfectly with TEXT40 (eg. in INVERSE, so the character cell has lit pixel in line with the HGR vertical line)
- an HGR pixel in the bottom-right (279,159) is full width, ie. 3DF7:40
  - and this aligns pefectly with TEXT40 (eg. in INVERSE)
  - currently in AppleWin this looks like a wider pixel
- a half-dot shifted HGR pixel in the bottom-right (279,159) is half width, ie. 3DF7:C0
  - currently in AppleWin this looks like a wider pixel

Other things I checked:
- DHGR dot in the top-right: yes this is visible (ie. the original point of this issue!)
- DHGR and TEXT80 is aligned (this was fixed above)

**Comment by tomcw** [comment]
>a half-dot shifted HGR pixel in the bottom-right (279,159) is half width, ie. 3DF7:C0 

AppleWin will draw a 3rd 14M/DHGR pixel for 3DF7:40
- For 1.27.4, all 3x 14M pixels are visible, so it's visible as an extra long pixel!
- For 1.27.2, only 2x 14M pixels are visible, but the 3rd off-screen pixel influences the colour (eg. in Color TV mode), so it starts green and finishes white.

I will change this for HGR, such that that last video byte (at hpos=64) clears `g_nLastColumnPixelNTSC` so that the end-of-line off-screen pixel is zero.

**Comment by tomcw** [comment]
NB. I have _reverted_ the fix for showing the 559th vertical column in [1.27.5](https://github.com/AppleWin/AppleWin/releases/tag/v1.27.5.0).

**Comment by landloafer** [comment]
Would overscan a few pixel outside this 560x384px help this issue?

**Comment by tomcw** [comment]
Perhaps a 561 pixel display is what is really required?

I want to back this up with evidence though using real h/w. Should be easy to verify by flicking between HGR and DHGR screens both with vertical lines at 0 and 279/559 respectively.

But from my comment [above](https://github.com/AppleWin/AppleWin/issues/555#issuecomment-394186908), in HGR mode there are only 560 DHGR 14M/pixels, since the pixel drawn at (279,159) with 3DF7:C0 is half-width.

So far I've been looking at AppleWin's NTSC s/w renderer (#557), but whenever I try to re-align either HGR (or DHGR) by one 14M/DHGR pixel, then the chroma signal is out of phase, leading to the wrong colours.

**Comment by tomcw** [comment]
On real //e h/w: flick between 40-col & 80-col INVERSE text to see the left/right edge differences:

```
PR#3: INVERSE: HOME

10 HOME
20 POKE 12*4096+12,0: GET A$
30 POKE 12*4096+13,0: GET A$
40 GOTO 20
```

**Comment by tomcw** [comment]
On my UK (unenhanced) //e, using monochrome video output, the 80-col display starts 7(?)x 14M pixels left of the 40-col screen:

Switch from 80-col to 40-col at bottom of screen:
![dsc_1136-s](https://user-images.githubusercontent.com/6696896/44002902-5849b3e4-9e42-11e8-9113-5fc3359c1868.png)

Switch from 40-col to 80-col in middle of screen:
![dsc_1137-s](https://user-images.githubusercontent.com/6696896/44002904-5a642358-9e42-11e8-80aa-f0693545f8dd.png)

**Comment by sicklittlemonkey** [comment]
Ok, I should probably mention this about now.

A while ago I did some experiments for Rebecca Bettencourt that she
requested at the bottom of this page:
http://www.kreativekorp.com/software/fonts/apple2.shtml
So I ran the test and sent her some pictures. I'll attach them.

She has some other archived info here:
http://www.kreativekorp.com/miscpages/a2info/videotheory.shtml
And that page links to her conclusions here:
https://docs.google.com/spreadsheets/d/1rKR6A_bVniSCtIP_rrv8QLWJdj4h6jEU1jJj0AebWwg/edit#gid=0
Note that the 14M shifted pixels (HIRES80 and TEXT80) start earlier than 7M
shifted ones (TEXT 40).

Re-reading the video section of Sather's Understanding the Apple //e, he
mentions the blanking period being extended, but doesn't draw attention to
this difference in mixed 40/80 TEXT/HIRES.
The only place I can find him referring to variation in the horizontal
blanking - and thus picture start - is page 8-26, column 2, second
paragraph, starting "The first time LDPS' falls".

Cheers,
Nick.

**Comment by tomcw** [comment]
A quick hack to check this:
- In GetFrameBufferBorderlessWidth() set uFrameBufferBorderlessW = 567 (was 560).
- At the end of each line, in updateVideoScannerAddress():
  - reset color phase to INITIAL_COLOR_PHASE+1 (was INITIAL_COLOR_PHASE).
  - if the video is 7M (ie. !80COL, or 80COL && MIXED && Y>=160) then start 7 pixels later, and advance the color phase by 7.
- For DHGR, in updateScreenDoubleHires80(), remove the `bits = (bits << 1)` 0.25 colour clock delay.

NB. There's other stuff to tidy up, eg. border needs clearing now, DGR colours need fixing, DHGR draws an extra pixel on right edge.

Also do we want this 7 pixel border? If nothing uses this 7-pixel shift, then we could just extend to  561 pixels and resolve this issue.

**Comment by tomcw** [comment]
Thanks Nick. Some useful info for sure! And good to see this backed up elsewhere. I need to (re-)read your Sather reference.

Any thoughts on the AppleWin's borderless video now potentially needing a 7-pixel border (left or right, depending on 7M or 14M)? Or do we just ignore this and extend to 561 pixels so that 7M and 14M are out by 1 pixel instead (I've also checked this and it looks/works fine too)?

**Comment by tomcw** [comment]
>It is possible to modify the starting phase and keep it in 560?

Unfortunately not. The best we can do is 1 pixel out (between 7M and 14M video modes).

One possibility would be to blit the 14M image back 1 pixel (and only do this for the top 160 lines in mixed mode), but this wouldn't work if a demo mixed 7M and 14M video modes in the same scanline(!). Arguably no demos currently do this, but just switching to 14M (eg. writing to $C00D from the monitor) will be switch 7M to 14M mid-display, so may result in an odd visual glitch as you write $C00D (or $C00C to go back to 7M).

The 1-pixel border (561-wide display) is the simplest fix for this issue, so this is my slight preference.

And if in the future someone writes a demo to exploit the 7-pixel border, then we can re-think this.

**Comment by landloafer** [comment]
I opt for 561px. No point keeping a black border all the time.

**Comment by sicklittlemonkey** [comment]
OK. I agree 1 pixel is visually preferable to 7.

Cheers,
Nick.

**Comment by tomcw** [comment]
Merged PR #577 to master at: 7d100a3.
Added tests for 40/80-col + mixed text & graphics video modes at: https://github.com/AppleWin/AppleWin-Test/commit/edb051b8ed3eab1b3aec1c96a1a9721b25ed989d

**Comment by tomcw** [comment]
FYI, the 1 pixel shift (ie. 1x 14M pixel) is barely visible. Here a screenshot showing mixed HGR+TEXT80 (where the TEXT80 part gets shifted by 1 pixel):

![applewin_screenshot_000000000](https://user-images.githubusercontent.com/6696896/45264826-983c6d80-b43a-11e8-8d5a-389ad8554fe0.png)

**Comment by tomcw** [comment]
...and now that I've done this work (and have a slightly better understanding of NTSC.cpp) I'm thinking why not:
- Either, at the end of each line, if it's 14M, then copy the line back 1 pixel
- Or better, at the start of each line, if it's 14M, then render 1 pixel earlier

Then I can revert to a 560-pixel wide display.
(All of this presupposes that there's no legitimate reason to mix 7M and 14M video modes on the same scanline!)

**Comment by tomcw** [comment]
> Or better, at the start of each line, if it's 14M, then render 1 pixel earlier

This seems to work fine (39f91c5).

**Comment by tomcw** [comment]
Same display as a few comments above - screenshot showing mixed HGR+TEXT80.
NB. The TEXT80 part is now _not_ shifted by 1 pixel.
![tests-various_000000000](https://user-images.githubusercontent.com/6696896/45265932-ea858a80-b44a-11e8-83c7-7757e21ad616.png)

**Comment by tomcw** [comment]
OK, I think this is now (finally!) fixed, so try [1.27.8](https://github.com/AppleWin/AppleWin/releases/download/v1.27.8.0/AppleWin1.27.8.0.zip) to test it.
